### PR TITLE
[TCES-24][MERGED]: Implement POST request for Job Postings

### PR DESCRIPTION
**Pull request by williamntlam**
# Description

This PR is for creating Job Postings via a POST request. This is for Issue #TCES-24 on linear.app.

## November 20th, 2024

Hi! I just finished implementing the POST request for Job Posts. Feel free to take a look and see if there's anything wrong. Thanks!

William.

Note: I tested it with Postman, but since the Job Posts UI isn't ready yet, it was hard to test whether req.user.id is successfully attaching the Job Posts to the correct owner so I just created a new User and attached it through changing the req.user.id to the id number of the newly created User to test it. It seems to work.

image

image

I attached a Postman test I did to test the POST request. It seems to work well!

## Checklist:
Confirm all required fields for Job Postings:
- [X] title
- [X] employer
- [X] location
- [X] hours_per_week
- [X] rate_of_pay_min
- [X] rate_of_pay_max
- [X] rate_of_pay_frequency
- [X] job_type
- [X] close_date
- [X] job_description

Job Posts POST requirements:

- [X] set state to be "Active"
- [X] Confirm application close date is not in the past
- [X] Fill in creator field with current user id

Confirm all required fields for Job Postings Drafts:
- [X] Confirm title (all other fields are optional)
- [X] set state to be "Draft"
- [X] Fill in creator field with current user id

## Remaining Tasks:

- [X] make a new controller file in server/src/controllers/job_posts
- [X] add the endpoint to server/src/routes/job_posts.js
- [X] make a call to the endpoint in client/src/utils/job_posts_api.js

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# Checklist

Please delete options that are not applicable.

- [X] My code follows the style guidelines of this project (required)
- [X] I have performed a self-review of my code (required)
- [X] I have linted and formatted my code (required)
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by williamntlam** [comment]
## November 20th, 2024

Hi! I just finished implementing the POST request for Job Posts. Feel free to take a look and see if there's anything wrong. Thanks!

image

image

I attached a Postman test I did to test the POST request. It seems to work well!

William.

**Comment by williamntlam** [comment]
Note: need to double check, but if the employer of a job post is deleted, are all the job posts deleted as well?

File: server/src/controllers/job_posts/addJobPost.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      const potentialEmployee = await Employee.findOne({
+        where: { name: employer },
+      });
+
+      if (!potentialEmployee) {
+        return res.status(400).json({
+          error: "Employee mentioned in the job posting does not exist. ",
+        });
+      }
+    }
```
Comment (04kash) [inline server/src/controllers/job_posts/addJobPost.js]: @miekoyao is it necessary for the employer mentioned in the job posting to also be included in the employer table?

**Review by 04kash — COMMENTED** [review]
lgtm! Just need @miekoyao to clarify on one thing.

File: server/src/controllers/job_posts/addJobPost.js (hunk 1 above)
Comment (04kash) [reply server/src/controllers/job_posts/addJobPost.js]: Confirmed with @miekoyao that this should be removed

**Comment by williamntlam** [comment]
Feel free to review my PR. Let me know if there is anything out of the ordinary!

William.

**Review by 04kash — APPROVED** [review]
Lgtm!

**Pull request merged**